lib/executor/vu_handle.go

```diff
@@ -77,52 +87,51 @@ func newStoppedVUHandle(
 
 	// TODO maybe move the repeating parts in a function?
 	vh.returnVU = func(v lib.InitializedVU) {
-		// Don't return the initialized VU back
 		vh.mutex.Lock()
-		select {
-		case <-vh.parentCtx.Done():
-			// we are done just ruturn the VU
+		// the returnVU could be called after those have reinitialize, don't nil them if that is the
+		// case, maybe just don't nil them
+		if vh.state != starting && vh.state != running {
 			vh.initVU = nil
 			vh.activeVU = nil
-			atomic.StoreInt32(&vh.change, 1)
-			vh.mutex.Unlock()
-			returnVU(v)
-		default:
-			select {
-			case <-vh.canStartIter:
-				// we can continue with itearting - lets not return the vu
-				vh.activateVU() // we still need to reactivate it to get the new context and cancel
-				atomic.StoreInt32(&vh.change, 1)
-				vh.mutex.Unlock()
-			default: // we actually have to return the vu
-				vh.initVU = nil
-				vh.activeVU = nil
-				atomic.StoreInt32(&vh.change, 1)
-				vh.mutex.Unlock()
-				returnVU(v)
-			}
 		}
+		vh.mutex.Unlock()
+		returnVU(v)
 	}
 
 	return vh
 }
 
+// start is one of the more complicated ones as it can race with toGracefulStop and has special
+// cases
+// state changes
+// toGracefulStop => running, we have raced with toGracefulStop if it was finished the state
+// would've been stoppped
+// starting, running => nothing, we have done what needs to be done
+// stopped, toHardStop => reactivate everything, it is either stopped or hardStop has uninitialized
+// some stuff and we need to reinitialize them
 func (vh *vuHandle) start() (err error) {
 	vh.mutex.Lock()
-	vh.logger.Debug("Start")
-	if vh.initVU == nil {
-		vh.initVU, err = vh.getVU()
-		if err != nil {
-			return err
-		}
-		vh.activateVU()
-		atomic.AddInt32(&vh.change, 1)
-	}
-	select {
-	case <-vh.canStartIter: // we are alrady started do nothing
-	default:
+	if vh.state == toGracefulStop { // No point returning the VU just continue as it was
 		close(vh.canStartIter)
+		vh.changeState(running)
+		vh.mutex.Unlock()
+		return
 	}
+
+	if vh.state == starting || vh.state == running {
+		vh.mutex.Unlock()
```

Comment:
Could we `defer vh.mutex.Unlock()` at the start to avoid repetition?

Reply:
yeah ... I start wondering the same as well .. it won't be possible in `runLoopsIfPossible`. but I would argue the slowdown from defer (which now is way smaller) is not such a problem for this methods ...

Reply:
fixed in 337fc21